### PR TITLE
colour sensor consistency

### PR DESCRIPTION
**Pull request by LucienMorey**
allow colour sensor enum to keep rolling tally of colour values when it sees a ball to increase its confidence over longer periods

### CODE REVIEW

File: components/indexer.py [hunk 1]
```diff
@@ -51,7 +66,7 @@ class Direction(Enum):
     _chimney_direction = Direction.OFF
     _cat_flap_is_open = False
 
-    last_colour = CargoColour.NONE
+    cargo_colour = CargoColour()
```
Comment (auscompgeek) [inline components/indexer.py:69]: Do not assign mutable objects to a class variable that are supposed to be defaults. Class variables are global state. Expect things to break badly once RobotPy has tests again.

**Review by auscompgeek — CHANGES_REQUESTED** [review]
If we're really going to go down the path of rolling totals of colour readings (which I'm still unconvinced is a good idea), then I'd like to see this API:

```
Indexer
+ get_colour() : CargoColour
+ reset_colour_reading() : None
```

where `CargoColour` is the existing enum with the `match_colour` method removed.

I won't dictate how the colour matching is implemented, but it should be considered an implementation detail to the controller, outside from that reset method (i.e. even the mere fact that it might be a separate object should be hidden from the controller).

**Comment by james-ward** [comment]
Hey @auscompgeek if you want to change the colour matching method that would be great. Yesterday we had some false identifications (which may have been caused by crap on the sensor) but we are worried that the current method of just looking at the last colour identification is unreliable. The last identification has the possibility of being one where the cargo is right on the limit of identification, meaning a false identification is more likely. Hence the move to keeping a running total - an unreliable reading at the end won't swamp the accumulated good readings.

If this is something you think we can do with the `rev.ColorMatch` then that's great too. But we need to fix this pronto, hence Lucien getting something written that seems to be working.

**Comment by auscompgeek** [comment]
> Hence the move to keeping a running total - an unreliable reading at the end won't swamp the accumulated good readings.

Did we consider taking multiple readings and taking the most common valid reading to deal with this instead?

File: components/indexer.py [hunk 2]
```diff
@@ -136,3 +130,23 @@ def run_tunnel_motor(self, direction: Direction) -> None:
...
+        if colour.red > 700 or colour.blue > 700:
+            self.red_total += colour.red
+            self.blue_total += colour.blue
+
+        if self.blue_total == 0 and self.red_total == 0:
+            return CargoColour.NONE
+        elif self.blue_total > self.red_total:
+            return CargoColour.BLUE
+        else:
+            return CargoColour.RED
```
Comment (auscompgeek) [inline components/indexer.py:148]: We'll get here if `red_total == blue_total`. Is this intended?

File: components/indexer.py [hunk 3]
```diff
@@ -100,7 +86,11 @@ def has_cargo_in_tunnel(self) -> bool:
 
     @feedback
     def last_cargo_was_opposition(self) -> bool:
-        return self.last_colour.is_opposition()
+        if wpilib.DriverStation.getAlliance() is wpilib.DriverStation.Alliance.kBlue:
+            return self.red_total > self.blue_total
+        elif wpilib.DriverStation.getAlliance() is wpilib.DriverStation.Alliance.kRed:
```
Comment (auscompgeek) [inline components/indexer.py:91]: pybind11 enums are weird :(
```suggestion
        if wpilib.DriverStation.getAlliance() == wpilib.DriverStation.Alliance.kBlue:
            return self.red_total > self.blue_total
        elif wpilib.DriverStation.getAlliance() == wpilib.DriverStation.Alliance.kRed:
```
although we could just do away with this altogether?

**Review by auscompgeek — APPROVED** [review]
LGTM, a couple of minor comments.

**Pull request merged**